crates/chia-client/src/request_map.rs

```diff
@@ -0,0 +1,65 @@
+use std::{collections::HashMap, sync::Arc};
+
+use chia_protocol::Message;
+use tokio::sync::{oneshot, Mutex, OwnedSemaphorePermit, Semaphore};
+
+#[derive(Debug)]
+pub struct Request {
+    sender: oneshot::Sender<Message>,
+    _permit: OwnedSemaphorePermit,
+}
+
+impl Request {
+    pub fn send(self, message: Message) {
+        self.sender.send(message).ok();
+    }
+}
+
+#[derive(Debug)]
+pub struct RequestMap {
+    items: Mutex<HashMap<u16, Request>>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl RequestMap {
+    pub fn new() -> Self {
+        Self {
+            items: Mutex::new(HashMap::new()),
+            semaphore: Arc::new(Semaphore::new(u16::MAX as usize)),
+        }
+    }
+
+    pub async fn insert(&self, sender: oneshot::Sender<Message>) -> u16 {
+        let permit = self
+            .semaphore
+            .clone()
+            .acquire_owned()
+            .await
+            .expect("semaphore closed");
+
+        let mut items = self.items.lock().await;
+
+        let mut index = None;
+
+        for i in 0..=u16::MAX {
+            if !items.contains_key(&i) {
+                index = Some(i);
+                break;
+            }
+        }
```

Comment:
one simple optimization over a linear search is to store a `free_slot: Option<u16>,` member, which is set every time you release a slot, and before this linear search, you could just see if that slot is available.

Another, more heavy weight, optimization would be to have a stack of all free slots that you push and pop from, but given that the common case is (likely) to have few messages in flight, that is probably not worth it.